File: py-appscript/trunk/test/test_appscriptcommands.py
#!/usr/local/bin/python

import unittest, commands
import appscript, aem, mactypes

class TC_appscriptNewApp(unittest.TestCase):

	def test_by_name(self):
		for name in [
				'/Applications/TextEdit.app',
				'Finder.app',
				'System Events'
		]:
			a = appscript.app(name)
			self.assertNotEqual(None, a)
			self.assertEqual(appscript.reference.Application, a.__class__)
			self.assertEqual(appscript.reference.Reference, a.name.__class__)
		
		self.assertEqual("app(u'/Applications/TextEdit.app')", str(appscript.app('TextEdit')))
		self.assertEqual("app(u'/Applications/TextEdit.app')", str(appscript.app(name='TextEdit')))
		
		self.assertRaises(appscript.ApplicationNotFoundError, appscript.app, '/non-existent/app')
		self.assertRaises(appscript.ApplicationNotFoundError, appscript.app, 'non-existent.app')
	
	
	def test_by_id(self):
		for name in [
				'com.apple.textedit',
				'com.apple.finder',
		]:
			a = appscript.app(id=name)
			self.assertNotEqual(None, a)
			self.assertEqual(appscript.reference.Application, a.__class__)
			self.assertEqual(appscript.reference.Reference, a.name.__class__)
		
		self.assertEqual("app(u'/Applications/TextEdit.app')", str(appscript.app(id='com.apple.textedit')))
		
		self.assertRaises(appscript.ApplicationNotFoundError, appscript.app, id='non.existent.app')
	

	def test_by_creator(self):
		a = appscript.app(creator='ttxt')
		self.assertEqual(appscript.reference.Application, a.__class__)
		self.assertEqual(appscript.reference.Reference, a.name.__class__)
		self.assertEqual("app(u'/Applications/TextEdit.app')", str(a))
		self.assertRaises(appscript.ApplicationNotFoundError, appscript.app, id='!self.$o')
	
	
	def test_by_pid(self):
		pid = int(commands.getoutput("top -l1 | grep Finder | awk '{ print $1 }'"))
		a = appscript.app(pid=pid)
		self.assertEqual(appscript.reference.Application, a.__class__)
		self.assertEqual(appscript.reference.Reference, a.name.__class__)
		self.assertEqual("app(pid=%i)" % pid, str(a))
		self.assertEqual('Finder', a.name.get())
	
	
	def test_by_aem_app(self):
		a = appscript.app(aemapp=aem.Application('/Applications/TextEdit.app'))
		self.assertEqual(appscript.reference.Application, a.__class__)
		self.assertEqual(appscript.reference.Reference, a.name.__class__)
		self.assertEqual("app(aemapp=aem.Application('/Applications/TextEdit.app'))", str(a))
	



class TC_appscriptCommands(unittest.TestCase):

	def setUp(self):
		self.te = appscript.app('TextEdit')
		self.f = appscript.app('Finder')
	
	
	def test_commands_1(self):
		self.assertEqual('TextEdit', self.te.name.get())
		d = self.te.make(new=appscript.k.document, at=appscript.app.documents.end, 
				with_properties={appscript.k.text:'test test_commands'})
		self.assertEqual(appscript.reference.Reference, d.__class__)
		d.text.end.make(new=appscript.k.word, with_data=' test2')
		self.assertEqual('test test_commands test2', d.text.get())
		self.assertEqual(unicode, 
				type(d.text.get(ignore=[appscript.k.diacriticals, appscript.k.punctuation, 
						appscript.k.whitespace, appscript.k.expansion], timeout=10)))
		self.assertEqual(None, d.get(waitreply=False))
		
		d.text.set(u'\xa9 M. Lef\xe8vre')
		self.assertEqual(u'\xa9 M. Lef\xe8vre', d.text.get())
		
		d.close(saving=appscript.k.no)
	
	
	def test_commands_2(self):
		d = self.te.make(new=appscript.k.document, at=self.te.documents.end)
		
		self.te.set(d.text, to= 'test1')
		self.assertEqual('test1', d.text.get())
		
		self.te.set(d.text, to= 'test2')
		self.te.make(new=appscript.k.word, at=appscript.app.documents[1].paragraphs.end, with_data=' test3')
		self.assertEqual('test2 test3', d.text.get())
		
		d.close(saving=appscript.k.no)
		
		self.assertRaises(appscript.CommandError, self.te.documents[10000].get)
		
		self.assertEqual(int, type(self.te.documents.count()))
		self.assertEqual(self.te.documents.count(), self.te.count(each=appscript.k.document))
	
	
	def test_commands_3(self):
		self.assertEqual('Finder', self.f.name.get())
		val = self.f.home.folders['Desktop'].get(resulttype=appscript.k.alias)
		self.assertEqual(mactypes.Alias, val.__class__)
		self.assertEqual(val, self.f.desktop.get(resulttype=appscript.k.alias))
		self.assertEqual(list, type(self.f.disks.get()))
		
		r = self.f.home.get()
		f = r.get(resulttype=appscript.k.file_ref)
		self.assertEqual(r, self.f.items[f].get())
		
		self.assertEqual(self.f.home.items.get(), self.f.home.items.get())
		self.assertNotEqual(self.f.disks['non-existent'], self.f.disks[1].get())
	
	
	def test_command_error(self):
		try:
			self.f.items[10000].get()
		except appscript.CommandError, e:
			self.assertEqual(-1728, int(e))
			self.assertEqual("Command failed:\n\t\tOSERROR: -1728\n\t\tMESSAGE: Can't get reference.\n\t\tOFFENDING OBJECT: app(u'/System/Library/CoreServices/Finder.app').items[10000]\n\t\tCOMMAND: app(u'/System/Library/CoreServices/Finder.app').items[10000].get()", str(e))
			self.assertEqual(aem.EventError, e.realerror.__class__)




if __name__ == '__main__':
	unittest.main()

File: rb-appscript/trunk/test/test_appscriptcommands.rb
#!/usr/bin/ruby -w

begin; require 'rubygems'; rescue LoadError; end

require 'test/unit'
require 'appscript'

# rb-appscript 0.5.0+ should no longer require the following kludge:
#class AS_SafeObject
#	def self.hide(name)
#	end
#end

class TC_AppscriptNewApp < Test::Unit::TestCase

	def test_by_name
		[
				'/Applications/TextEdit.app',
				'Finder.app',
				'System Events'
		].each do |name|
			a = Appscript.app(name)
			assert_not_nil(a)
			assert_instance_of(Appscript::Application, a)
			assert_instance_of(Appscript::Reference, a.name)
		end
		assert_equal('app("/Applications/TextEdit.app")', Appscript.app('TextEdit').to_s)
		assert_equal('app("/Applications/TextEdit.app")', Appscript.app.by_name('TextEdit').to_s)
		
		assert_raises(Appscript::ApplicationNotFoundError) { Appscript.app('/non-existent/app') }
		assert_raises(Appscript::ApplicationNotFoundError) { Appscript.app('non-existent.app') }
	end
	
	def test_by_id
		[
				'com.apple.textedit',
				'com.apple.finder',
		].each do |name|
			a = Appscript.app.by_id(name)
			assert_not_nil(a)
			assert_instance_of(Appscript::Application, a)
			assert_instance_of(Appscript::Reference, a.name)
		end
		assert_equal('app("/Applications/TextEdit.app")', Appscript.app.by_id('com.apple.textedit').to_s)
		
		assert_raises(Appscript::ApplicationNotFoundError) { Appscript.app.by_id('non.existent.app') }
	end

	def test_by_creator
		a = Appscript.app.by_creator('ttxt')
		assert_instance_of(Appscript::Reference, a.name)
		assert_equal('app("/Applications/TextEdit.app")', a.to_s)
		assert_raises(Appscript::ApplicationNotFoundError) { Appscript.app.by_id('!@$o') }
	end
	
	def test_by_pid
		pid = `top -l1 | grep Finder | awk '{ print $1 }'`.to_i # note: this line will return a bad result if more than one user is logged in
		a = Appscript.app.by_pid(pid)
		assert_instance_of(Appscript::Reference, a.name)
		assert_equal("app.by_pid(#{pid})", a.to_s)
		assert_equal('Finder', a.name.get)
	end
	
	def test_by_aem_app
		a = Appscript.app.by_aem_app(AEM::Application.by_path('/Applications/TextEdit.app'))
		assert_instance_of(Appscript::Reference, a.name)
		assert_equal('app.by_aem_app(AEM::Application.by_path("/Applications/TextEdit.app"))', a.to_s)
	end
end


class TC_AppscriptCommands < Test::Unit::TestCase

	def setup
		@te = Appscript.app('TextEdit')
		@f = Appscript.app('Finder')
	end
	
	def test_commands_1
		assert_equal('TextEdit', @te.name.get)
		d = @te.make(:new=>:document, :with_properties=>{:text=>'test test_commands'})
		assert_instance_of(Appscript::Reference, d)
		d.text.end.make(:new=>:word, :with_data=>' test2')
		assert_equal('test test_commands test2', d.text.get)
		assert_instance_of(String, 
				d.text.get(:ignore=>[:diacriticals, :punctuation, :whitespace, :expansion], :timeout=>10))
		assert_nil(d.get(:wait_reply=>false))
		
		
		# test Ruby 1.9+ String Encoding support
		version, sub_version = RUBY_VERSION.split('.').collect {|n| n.to_i} [0, 2]
		if version >= 1 and sub_version >= 9
			
			print "(check Encoding support)"
			s = "\302\251 M. Lef\303\250vre"
			s.force_encoding('utf-8')
			d.text.set(s)
			assert_equal(s, d.text.get)
		
			@te.AS_app_data.use_ascii_8bit
		end
		
		d.text.set("\302\251 M. Lef\303\250vre")
		assert_equal("\302\251 M. Lef\303\250vre", d.text.get)
		
		d.close(:saving=>:no)
	end
	
	def test_commands_2
		d = @te.make(:new=>:document, :at=>@te.documents.end)
		
		@te.set(d.text, :to=> 'test1')
		assert_equal('test1', d.text.get)
		
		@te.set(d.text, :to=> 'test2')
		@te.make(:new=>:word, :at=>Appscript.app.documents[1].paragraphs.end, :with_data=>' test3')
		assert_equal('test2 test3', d.text.get)
		
		d.close(:saving=>:no)
		
		assert_raises(Appscript::CommandError) { @te.documents[10000].get }
		
		assert_instance_of(Fixnum, @te.documents.count)
		assert_equal(@te.documents.count, @te.count(:each=>:document))
	end
	
	def test_commands_3
		assert_equal('Finder', @f.name.get)
		val = @f.home.folders['Desktop'].get(:result_type=>:alias)
		assert_instance_of(MacTypes::Alias, val)
		assert_equal(val, @f.desktop.get(:result_type=>:alias))
		assert_instance_of(Array, @f.disks.get)
		
		r = @f.home.get
		f = r.get(:result_type=>:file_ref)
		assert_equal(r, @f.items[f].get)
		
		assert_equal(@f.home.items.get, @f.home.items.get)
		assert_not_equal(@f.disks['non-existent'], @f.disks[1].get)
	end
	
	def test_command_error
		begin
			@f.items[10000].get
		rescue Appscript::CommandError => e
			assert_equal(-1728, e.to_i)
			assert_equal("CommandError\n\t\tOSERROR: -1728\n\t\tMESSAGE: Can't get reference.\n\t\tOFFENDING OBJECT: app(\"/System/Library/CoreServices/Finder.app\").items[10000]\n\t\tCOMMAND: app(\"/System/Library/CoreServices/Finder.app\").items[10000].get()\n", e.to_s)
			assert_instance_of(AEM::EventError, e.real_error)
		end
	end
end

